src/ILCompiler.CppCodeGen/src/CppCodeGen/CppWriter.cs

```diff
@@ -232,6 +232,18 @@ public void AppendCppMethodCallParamList(CppGenerationBuffer sb, MethodDesc meth
 
             for (int i = 0; i < argCount; i++)
             {
+                if(i == 0 && unbox)
+                {
+                    // Unboxing stubs only valid for non-static methods
+                    System.Diagnostics.Debug.Assert(hasThis);
+
+                    var thisType = method.OwningType;
+                    if (thisType.IsValueType)
```

Comment:
I think that this should be assert instead as well - unboxing stubs are only valid for value types.